crates/task-impls/src/quorum_vote.rs

```diff
@@ -70,12 +74,169 @@ struct VoteDependencyHandle<TYPES: NodeType, I: NodeImplementation<TYPES>> {
+        //         QuorumData {
+        //             leaf_commit: proposed_leaf.commit(),
+        //         },
+        //         view,
+        //         &public_key,
+        //         &self.private_key
+        //     ) {
+        //         GeneralConsensusMessage::<TYPES>::Vote(vote)
+        //     } else {
+        //         error!("Unable to sign quorum vote!");
+        //         return false;
+        //     }
+        // } else {
+        //     error!(
+        //         "Invalid DAC in proposal! Skipping proposal. {:?} cur view is: {:?}",
+        //         cert, cur_view
+        //     );
+        //     return false;
+        // };
+
+        // Now that we've rounded everyone up, we need to update the shared state and broadcast our events.
+        // We will defer broadcast until all states are updated to avoid holding onto the lock during a network call.
+        let mut consensus_writer = self.consensus.write().await;
+
+        let view = View {
+            view_inner: ViewInner::Leaf {
+                leaf: proposed_leaf.commit(),
+                state: Arc::clone(&state),
+                delta: Some(Arc::clone(&delta)),
+            },
+        };
+        consensus_writer.update_validated_state_map(proposed_leaf.view_number(), view.clone());
+        consensus_writer.update_saved_leaves(proposed_leaf.clone());
+
+        // Kick back our updated structures for downstream usage.
+        let new_leaves = consensus_writer.saved_leaves().clone();
+        let new_state = consensus_writer.validated_state_map().clone();
+        drop(consensus_writer);
+
+        // Broadcast now that the lock is dropped.
+        broadcast_event(
+            HotShotEvent::ValidatedStateUpdated(self.view_number, view).into(),
+            &self.sender,
+        )
+        .await;
+
+        // Send the new state up to the sequencer.
+        self.storage
+            .write()
+            .await
+            .update_undecided_state(new_leaves, new_state)
+            .await?;
+
+        Ok(())
+    }
+
+    /// Submits the `QuorumVoteSend` event if all the dependencies are met.
+    async fn submit_vote(
+        &self,
+        leaf: Leaf<TYPES>,
+        vid_share: Proposal<TYPES, VidDisperseShare<TYPES>>,
+    ) -> Result<()> {
+        ensure!(
+            self.quorum_membership.has_stake(&self.public_key),
+            format!(
+                "We were not chosen for quorum committee on {:?}",
+                self.view_number
+            ),
+        );
+
+        // Create and send the vote.
+        let vote = QuorumVote::<TYPES>::create_signed_vote(
+            QuorumData {
+                leaf_commit: leaf.commit(),
+            },
+            self.view_number,
+            &self.public_key,
+            &self.private_key,
+        )
+        .context("Failed to sign vote")?;
+        let message = GeneralConsensusMessage::<TYPES>::Vote(vote);
+        if let GeneralConsensusMessage::Vote(vote) = message {
```

Comment:
It's rust's type system being a pain. First we create the signed vote, then we make the vote message, then we basically have to say "hey, I *promise* this is a vote" because we're using an enum.

Reply:
No you're right good catch! 

```diff
            &self.private_key,
         )
         .context("Failed to sign vote")?;
-        let message = GeneralConsensusMessage::<TYPES>::Vote(vote);
-        if let GeneralConsensusMessage::Vote(vote) = message {
-            debug!(
-                "Sending vote to next quorum leader {:?}",
-                vote.view_number() + 1
-            );
-            // Add to the storage.
-            self.storage
-                .write()
-                .await
-                .append_vid(&vid_share)
-                .await
-                .context("Failed to store VID share")?;
-            broadcast_event(Arc::new(HotShotEvent::QuorumVoteSend(vote)), &self.sender).await;
-        }
+        debug!(
+            "Sending vote to next quorum leader {:?}",
+            vote.view_number() + 1
+        );
+        // Add to the storage.
+        self.storage
+            .write()
+            .await
+            .append_vid(&vid_share)
+            .await
+            .context("Failed to store VID share")?;
+        broadcast_event(Arc::new(HotShotEvent::QuorumVoteSend(vote)), &self.sender).await;
 
         Ok(())
     }
```